prov/efa/src/efa_msg.c

```diff
@@ -0,0 +1,378 @@
+/* SPDX-License-Identifier: BSD-2-Clause OR GPL-2.0-only */
+/* SPDX-FileCopyrightText: Copyright (c) 2013-2015 Intel Corporation, Inc.  All rights reserved. */
+/* SPDX-FileCopyrightText: Copyright Amazon.com, Inc. or its affiliates. All rights reserved. */
+
+#include "config.h"
+
+
+#include "ofi.h"
+#include "ofi_enosys.h"
+#include "ofi_iov.h"
+
+#include "efa.h"
+#include "efa_av.h"
+
+#include "efa_tp.h"
+
+#define EFA_SETUP_IOV(iov, buf, len)           \
+	do {                                   \
+		iov.iov_base = (void *)buf;    \
+		iov.iov_len = (size_t)len;     \
+	} while (0)
+
+#define EFA_SETUP_MSG(msg, iov, _desc, count, _addr, _context, _data)    \
+	do {                                                             \
+		msg.msg_iov = (const struct iovec *)iov;                 \
+		msg.desc = (void **)_desc;                               \
+		msg.iov_count = (size_t)count;                           \
+		msg.addr = (fi_addr_t)_addr;                             \
+		msg.context = (void *)_context;                          \
+		msg.data = (uint32_t)_data;                              \
+	} while (0)
+
+#ifndef EFA_MSG_DUMP
+static inline void dump_msg(const struct fi_msg *msg, const char *context) {}
+#else
+#define DUMP_IOV(i, iov, desc) \
+	EFA_DBG(FI_LOG_EP_DATA, \
+		"\t{ iov[%d] = { base = %p, buff = \"%s\", len = %zu }, desc = %p },\n", \
+		i, iov.iov_base, (char *)iov.iov_base, iov.iov_len, (desc ? desc[i] : NULL))
+
+static inline void dump_msg(const struct fi_msg *msg, const char *context)
+{
+	int i;
+
+	EFA_DBG(FI_LOG_EP_DATA, "%s: { data = %u, addr = %" PRIu64 ", iov_count = %zu, [\n",
+		context, (unsigned)msg->data, msg->addr, msg->iov_count);
+	for (i = 0; i < msg->iov_count; ++i)
+		DUMP_IOV(i, msg->msg_iov[i], msg->desc);
+	EFA_DBG(FI_LOG_EP_DATA, " ] }\n");
+}
+#endif /* EFA_MSG_DUMP */
+
+/**
+ * @brief post receive buffer to EFA device via ibv_post_recv
+ *
+ * @param[in]	base_ep	endpoint
+ * @param[in]	msg	libfabric message
+ * @param[in]	flags	libfabric flags, currently only FI_MORE is supported.
+ * @reutrn	On Success, return 0
+ * 		On failure, return negative libfabric error code
+ */
+static inline ssize_t efa_post_recv(struct efa_base_ep *base_ep, const struct fi_msg *msg, uint64_t flags)
+{
+	struct efa_mr *efa_mr;
+	struct efa_qp *qp = base_ep->qp;
+	struct ibv_recv_wr *bad_wr;
+	struct ibv_recv_wr *wr;
+	uintptr_t addr;
+	ssize_t err, post_recv_err;
+	size_t i, wr_index = base_ep->recv_wr_index;
+
+	if (wr_index >= base_ep->info->rx_attr->size) {
+		EFA_WARN(FI_LOG_EP_DATA,
```

Comment:
Make the log in info level. Also you can just `return -FI_EAGAIN` without assigning it to `err`.

Reply:
Correct my last comment. I think in this case we need to flush the recv, we should goto `out_err`